src/Text/LLVM/PP.hs

```diff
@@ -299,34 +299,21 @@ ppTypeDecl td = ppIdent (typeName td) <+> char '='
 
 ppGlobal :: Fmt Global
 ppGlobal g = ppSymbol (globalSym g) <+> char '='
-         <+> ppTheGlobalAttrs (globalAttrs g)
+         <+> ppGlobalAttrs (isJust $ globalValue g) (globalAttrs g)
          <+> ppType (globalType g) <+> ppMaybe ppValue (globalValue g)
           <> ppAlign (globalAlign g)
           <> ppAttachedMetadata (Map.toList (globalMetadata g))
-  where
-  isStruct | Just (ValStruct {}) <- globalValue g = True
-           | otherwise = False
-  ppTheGlobalAttrs | isStruct = ppStructGlobalAttrs
-                    | otherwise = ppGlobalAttrs
 
-ppGlobalAttrs :: Fmt GlobalAttrs
-ppGlobalAttrs ga
+ppGlobalAttrs :: Bool -> Fmt GlobalAttrs
+ppGlobalAttrs hasValue ga
     -- LLVM 3.8 does not emit or parse linkage information w/ hidden visibility
     | Just HiddenVisibility <- gaVisibility ga =
             ppVisibility HiddenVisibility <+> constant
-    | otherwise = ppMaybe ppLinkage (gaLinkage ga) <+> ppMaybe ppVisibility (gaVisibility ga) <+> constant
-  where
-  constant | gaConstant ga = "constant"
-           | otherwise     = "global"
-
-ppStructGlobalAttrs :: Fmt GlobalAttrs
-ppStructGlobalAttrs ga
-    -- LLVM 3.8 does not emit or parse external linkage for
-    -- global structs
-    | Just External <- gaLinkage ga,
-      Just DefaultVisibility <- gaVisibility ga
-        = constant
-    | otherwise = ppGlobalAttrs ga
+    | Just External <- gaLinkage ga
+    , Just DefaultVisibility <- gaVisibility ga
```

Comment:
If I understand this code correctly, this will only omit things if the global is both `external` _and_ `default`. What happens if it is `external` but not `default`, or `default` but not `external`—will `llvm-as` accept it then? (The phrasing "either `external` or `default`" on #128 led me to believe that that was not the case, but perhaps I misread that.)

Reply:
Empirically, issue 128 is correct: neither is accepted, although issue #128 doesn't cover the case where there is no initial value.  The first commit I made on this PR handled 128 directly, but when I discovered that failed when it was truly an external value, I made adjustments, noticed that the result was nearly identical to the "structure" version, so I consolidated both here.  That may not be the correct resulting behavior, but so far I have never seen `external` without `default` (and vice-versa) based on empirical testing, and there is no other test framework to validate this functionality right now.

By empirically, I mean that I have tests (for llvm-pretty-bc-parser) that cover all 6 cases that I can think of:
* uniniitialized scalar
* initialized scalar
* uninitialized structure
* initialized structure
* external scalar
* external structure

These tests pass on my local copy of llvm-pretty-bc-parser/llvm-pretty that contain the aggregation of PRs 228 and 252 for the former and PRs 125, 127, 129, and 130 for the latter (all of which are needed to get things successful for LLVM 10 and 11).  I should have mentioned these and the fact that they are passing and my intent to commit them once this batch of PRs lands.

Reply:
OK, good to know. It would be worth leaving a comment next to this code explaining that the logic here is the result of empirical testing about places where `external`/`default` are permitted.